server/src/main/java/org/opensearch/common/util/LongRHHash.java

```diff
@@ -0,0 +1,242 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.util;
+
+import com.carrotsearch.hppc.BitMixer;
+import org.opensearch.common.lease.Releasable;
+
+import java.util.Arrays;
+
+/**
+ * Specialized hash table implementation that maps (primitive) long to long.
+ *
+ * It organizes itself by moving keys around dynamically in order to reduce the longest
+ * probe sequence length (PSL). It also optimizes lookups for recently accessed keys, which
+ * makes it very useful for aggregations where keys are correlated across consecutive hits.
+ *
+ * This class is not thread-safe.
+ *
+ * @opensearch.internal
+ */
+public class LongRHHash implements Releasable {
+    private static final long DEFAULT_CAPACITY = 1;
+    private static final float DEFAULT_LOAD_FACTOR = 0.6f;
+    private static final int DEFAULT_HINTS = 256;
+
+    /**
+     * Utility class to allocate big arrays.
+     */
+    private final BigArrays bigArrays;
+
+    /**
+     * Maximum load factor after which the capacity is doubled.
+     */
+    private final float loadFactor;
+
+    /**
+     * Current capacity of the hash table.
+     * This must always be a power of two since we use a bit-mask to calculate the slot.
+     */
+    private long capacity;
+
+    /**
+     * Bit-mask used to calculate the slot (i.e. the index) after hashing a key.
+     */
+    private long mask;
+
+    /**
+     * Current size of the hash table.
+     */
+    private long size;
+
+    /**
+     * Threshold at which the hash table needs to be resized.
+     */
+    private long grow;
+
+    /**
+     * Lookup table for ordinals.
+     */
+    private LongArray ords;
+
+    /**
+     * Lookup table for keys.
+     */
+    private LongArray keys;
+
+    /**
+     * Lookup table for "recent" key hints.
+     */
+    private final long[] hints;
+
+    /**
+     * Bit-mask used to calculate the slot in the hints' table.
+     */
+    private final int hintMask;
+
+    public LongRHHash(BigArrays bigArrays) {
+        this(DEFAULT_CAPACITY, DEFAULT_LOAD_FACTOR, DEFAULT_HINTS, bigArrays);
+    }
+
+    public LongRHHash(long capacity, float loadFactor, int numHints, BigArrays bigArrays) {
```

Comment:
Can we add validations on capacity, loadFactor and numHints?

Reply:
Ack!